csrc/alias_analysis.cpp

```diff
@@ -113,114 +128,123 @@ std::pair<bool, std::optional<bool>> mergeContiguity(
   return {false, std::nullopt};
 }
 
-void AliasFinder::handle(const ViewOp* view) {
-  TensorView* in = view->in();
-  TensorView* out = view->out();
-
-  const std::vector<IterDomain*>& in_rfactor = in->getMaybeRFactorDomain();
-  const std::vector<IterDomain*>& out_root = out->getRootDomain();
-  const std::vector<IterDomain*>& out_rfactor = out->getMaybeRFactorDomain();
-
-  Layout in_layout = analysis_.preferredLayout(in);
-  if (!ir_utils::computePermutation(in_rfactor, in_layout.allocation_domain)
+/*static*/ std::optional<Layout> AliasFinder::mapInLayoutToOutRoot(
+    const Layout& preferred_in_layout,
+    TensorView* in,
+    TensorView* out) {
+  if (!ir_utils::computePermutation(
+           in->getMaybeRFactorDomain(), preferred_in_layout.allocation_domain)
            .has_value()) {
-    // Give up when `in`'s allocation domain is not an rfactor permutation.
-    return;
+    // Give up when `in`'s allocation domain is not an rfactor permutation. As
+    // an extension, we could map in_alloc to in_rfactor and apply the inverse
+    // mapping to out_root.
+    return std::nullopt;
   }
 
   std::unordered_map<IterDomain*, IterDomain*> in_rfactor_to_out_root =
       PairwiseRootDomainMap(in, out).mapProducerToConsumer();
-  std::unordered_map<IterDomain*, IterDomain*> out_root_to_in_rfactor =
-      PairwiseRootDomainMap(in, out).mapConsumerToProducer();
+
+  Layout preferred_out_layout;
+  for (const auto i : c10::irange(preferred_in_layout.size())) {
+    IterDomain* in_alloc_id = preferred_in_layout.allocation_domain[i];
+    IterDomain* out_root_id = getOrDefault(in_rfactor_to_out_root, in_alloc_id);
+    if (out_root_id == nullptr) {
+      // This can happen when in_alloc_id is of type reduction or squeezed out.
+      continue;
+    }
+    preferred_out_layout.allocation_domain.push_back(out_root_id);
+    preferred_out_layout.contiguity.push_back(
+        preferred_in_layout.contiguity[i]);
+  }
+  return preferred_out_layout;
+}
+
+void AliasFinder::handle(const ViewOp* view) {
+  TensorView* in = view->in();
+  TensorView* out = view->out();
 
   // Collect the allocation order of `in`'s rfactor domain and thus `out`'s root
   // domain.
+  std::optional<Layout> out_root_layout =
+      mapInLayoutToOutRoot(analysis_.preferredLayout(in), in, out);
+  if (!out_root_layout.has_value()) {
+    return;
+  }
+
   LinkedHashMap<IterDomain*, std::optional<bool>> allocation_to_contiguity;
-  for (const auto i : c10::irange(in_layout.allocation_domain.size())) {
-    IterDomain* in_allocation_id = in_layout.allocation_domain[i];
-    if (in_allocation_id->isReduction()) {
-      // Reduction IterDomains won't appear in `out_root`.
-      continue;
-    }
+  for (const auto i : c10::irange(out_root_layout->size())) {
     allocation_to_contiguity.pushBack(
-        in_allocation_id, in_layout.contiguity[i]);
```

Comment:
Context: to work around #1174, `allocation_to_contiguity` may contain IterDomains both from the input TV and from the output. This however is unnecessary. The new code only puts output IterDomains in `allocation_to_contiguity` and looks up `out_root_to_in_rfactor` for expanded extents. It has the equivalent behavior but is more intuitive and doesn't need the `in_rfactor_to_out_root` map.